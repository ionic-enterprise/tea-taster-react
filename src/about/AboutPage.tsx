import { IonContent, IonHeader, IonPage, IonTitle, IonToolbar } from '@ionic/react';

const AboutPage: React.FC = () => {
  return (
    <IonPage>
      <IonHeader>
        <IonToolbar>
          <IonTitle>About Page</IonTitle>
        </IonToolbar>
      </IonHeader>
      <IonContent className="main-content"></IonContent>
    </IonPage>
  );
};
export default AboutPage;
