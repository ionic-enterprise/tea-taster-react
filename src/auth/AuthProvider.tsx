import { ReactNode, createContext, useContext, useEffect, useState } from 'react';
import * as auth from '../api/auth-api';
import { getSession } from '../api/session-api';
import { IonSpinner } from '@ionic/react';

type Props = { children?: ReactNode };

type Context = {
  isAuthenticated?: boolean;
  login: (email: string, password: string) => Promise<boolean>;
  logout: () => Promise<void>;
};

const AuthContext = createContext<Context | undefined>(undefined);
const AuthProvider = ({ children }: Props) => {
  const [isAuthenticated, setIsAuthenticated] = useState<boolean | undefined>(undefined);

  useEffect(() => {
    getSession().then((res) => setIsAuthenticated(!!res));
  }, []);

  const login = async (email: string, password: string): Promise<boolean> => {
    const isAuthenticated = await auth.login(email, password);
    setIsAuthenticated(isAuthenticated);
    return isAuthenticated;
  };

  const logout = async () => {
    await auth.logout();
    setIsAuthenticated(false);
  };

  return (
    <AuthContext.Provider value={{ isAuthenticated, login, logout }}>
      {isAuthenticated === undefined ? <IonSpinner /> : children}
    </AuthContext.Provider>
  );
};

export const useAuth = () => {
  const context = useContext(AuthContext);
  if (context === undefined) throw new Error('useAuth must be used within AuthProvider');
  return context;
};

export default AuthProvider;
