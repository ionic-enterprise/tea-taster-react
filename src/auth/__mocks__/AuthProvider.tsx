import { vi } from 'vitest';

const login = vi.fn().mockResolvedValue(undefined);
const logout = vi.fn().mockResolvedValue(undefined);

export const useAuth = () => ({ login, logout });
