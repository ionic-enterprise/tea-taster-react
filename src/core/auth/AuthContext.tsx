import { LockEvent } from '@ionic-enterprise/identity-vault';
import React, { createContext, useEffect, useReducer, useState } from 'react';
import { Session } from '../models';
import { SessionVault } from '../session-vault/SessionVault';

interface AuthState {
  session?: Session;
  loading: boolean;
  error: string;
}

const initialState: AuthState = {
  session: undefined,
  loading: false,
  error: '',
};

export type AuthAction =
  | { type: 'CLEAR_SESSION' }
  | { type: 'RESTORE_SESSION'; session: Session }
  | { type: 'LOGIN' }
  | { type: 'LOGIN_SUCCESS'; session: Session }
  | { type: 'LOGIN_FAILURE'; error: string }
  | { type: 'LOGOUT' }
  | { type: 'LOGOUT_SUCCESS' }
  | { type: 'LOGOUT_FAILURE'; error: string };

const reducer = (
  state: AuthState = initialState,
  action: AuthAction,
): AuthState => {
  switch (action.type) {
    case 'CLEAR_SESSION':
      return { ...state, session: undefined };
    case 'RESTORE_SESSION':
      return { ...state, session: action.session };
    case 'LOGIN':
      return { ...state, loading: true, error: '' };
    case 'LOGIN_SUCCESS':
      return { ...state, loading: false, session: action.session };
    case 'LOGIN_FAILURE':
      return { ...state, loading: false, error: action.error };
    case 'LOGOUT':
      return { ...state, loading: true, error: '' };
    case 'LOGOUT_SUCCESS':
      return { ...state, loading: false, session: undefined };
    case 'LOGOUT_FAILURE':
      return { ...state, loading: false, error: action.error };
    default:
      return state;
  }
};

export const AuthContext = createContext<{
  state: typeof initialState;
  dispatch: (action: AuthAction) => void;
  vault: SessionVault;
}>({
  state: initialState,
  dispatch: () => {},
  vault: SessionVault.getInstance(),
});

export const AuthProvider: React.FC<{ displayPasscodeRequest: () => void }> = ({
  displayPasscodeRequest,
  children,
}) => {
  const vault = SessionVault.getInstance();

  // Pass a function that displays the modal.

  const [initializing, setInitializing] = useState<boolean>(true);
  const [state, dispatch] = useReducer(reducer, initialState);

  useEffect(() => {
    (async () => {
      const session = await vault.restoreSession();
      if (!session) return setInitializing(false);
      dispatch({ type: 'RESTORE_SESSION', session });
      return setInitializing(false);
    })();
  }, [vault]);

  ///TODO: ADD LOCK/UNLOCK EVENTS

  vault.onVaultLocked = (event: LockEvent) => {
    dispatch({ type: 'CLEAR_SESSION' });
  };

  vault.onPasscodeRequest = async () => {
    return new Promise((res, reject) => {
      //displayPasscodeRequest(res, reject);
      // return with pin
    });
  };

  ///TODO: PROVIDE PASSCODE MECHANISM

  return (
    <AuthContext.Provider value={{ state, dispatch, vault }}>
      {initializing ? <div>Loading...</div> : children}
    </AuthContext.Provider>
  );
};
