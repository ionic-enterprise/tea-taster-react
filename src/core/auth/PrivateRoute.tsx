import React from 'react';
import { Route, Redirect, RouteProps } from 'react-router';
import { useSessionVault } from '../vault';
import { useSession } from './useSession';

interface PrivateRouteProps extends RouteProps {
  component: React.ComponentType<any>;
}

export const PrivateRoute: React.FC<PrivateRouteProps> = ({ component: Component, ...rest }) => {
  const { isAuthenticated } = useSession();
  const { isLocked } = useSessionVault();

  return (
    <Route
      {...rest}
      render={(props) => (!isLocked && isAuthenticated ? <Component {...props} /> : <Redirect to="/login" />)}
    />
  );
};
