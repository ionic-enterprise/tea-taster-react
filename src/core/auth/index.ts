export * from './SessionContext';
export * from './PrivateRoute';
export * from './useAuthentication';
export * from './useAuthInterceptor';
