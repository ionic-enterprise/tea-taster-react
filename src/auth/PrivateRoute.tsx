import { ReactNode } from 'react';
import { Redirect } from 'react-router';
import { useAuth } from './AuthProvider';

type Props = { children?: ReactNode };

export const PrivateRoute = ({ children }: Props) => {
  const { isAuthenticated } = useAuth();

  if (!isAuthenticated) return <Redirect to="/login" />;

  return <>{children}</>;
};
