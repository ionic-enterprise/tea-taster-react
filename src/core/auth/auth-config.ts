/* eslint-disable import/no-anonymous-default-export */
import { isPlatform } from '@ionic/react';
import { IonicAuthOptions } from '@ionic-enterprise/auth';

export default (): { config: IonicAuthOptions } => {
  const baseConfig = {
    clientID: '4geagm2idmq87fii15dq9toild',
    discoveryUrl: 'https://cognito-idp.us-east-2.amazonaws.com/us-east-2_YU8VQe29z/.well-known/openid-configuration',
    clientSecret: '124dch1p6824ppuef8o71unk14d4pt3p5hnntofvu21i2m960r1g',
    scope: 'openid email profile',
    audience: '',
  };

  const mobileAuthConfig: IonicAuthOptions = {
    ...baseConfig,
    authConfig: 'cognito',
    redirectUri: 'msauth://login',
    logoutUrl: 'msauth://login',
    platform: 'cordova',
    iosWebView: 'private',
    androidToolbarColor: 'Red',
  };

  const webAuthConfig: IonicAuthOptions = {
    ...baseConfig,
    authConfig: 'cognito',
    redirectUri: 'http://localhost:8100/login',
    logoutUrl: 'http://localhost:8100/login',
    platform: 'web',
  };

  const config = isPlatform('hybrid') ? mobileAuthConfig : webAuthConfig;

  return { config };
};
