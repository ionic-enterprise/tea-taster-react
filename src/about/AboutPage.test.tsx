import { Mock, vi } from 'vitest';
import { fireEvent, render, waitFor, screen } from '@testing-library/react';
import AboutPage from './AboutPage';
import { useAuth } from '../auth/AuthProvider';
import { useHistory } from 'react-router';

vi.mock('../auth/AuthProvider');
vi.mock('react-router');

describe('<AboutPage />', () => {
  it('renders without crashing', () => {
    const { baseElement } = render(<AboutPage />);
    expect(baseElement).toBeDefined();
  });

  it('renders consistently', () => {
    const { asFragment } = render(<AboutPage />);
    expect(asFragment()).toMatchSnapshot();
  });

  describe('sign out button', () => {
    it('performs a logout when clicked', async () => {
      const { logout } = useAuth();
      render(<AboutPage />);
      const button = screen.getByTestId('logout-button');
      fireEvent.click(button);
      await waitFor(() => expect(logout).toHaveBeenCalledTimes(1));
    });

    it('navigates to the login page', async () => {
      const history = useHistory();
      render(<AboutPage />);
      const button = screen.getByTestId('logout-button');
      fireEvent.click(button);
      await waitFor(() => expect(history.replace).toHaveBeenCalledTimes(1));
      expect(history.replace).toHaveBeenCalledWith('/login');
    });
  });
});
