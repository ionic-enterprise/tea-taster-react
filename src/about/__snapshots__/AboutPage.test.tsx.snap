// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AboutPage /> renders consistently 1`] = `
<DocumentFragment>
  <div
    class="ion-page"
  >
    <ion-header
      class="md header-md header-collapse-none"
      role="banner"
    >
      <!---->
      <ion-toolbar
        class="toolbar-title-default md in-toolbar"
      >
        <ion-title
          class="md title-default"
        >
          About Tea Taster
        </ion-title>
        <ion-buttons
          class="buttons-last-slot sc-ion-buttons-md-h sc-ion-buttons-md-s md"
          slot="end"
        >
          <ion-button
            class="md button button-clear in-toolbar button-has-icon-only ion-activatable ion-focusable"
            data-testid="logout-button"
          >
            <ion-icon
              class="md"
              icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Log Out</title><path d='M304 336v40a40 40 0 01-40 40H104a40 40 0 01-40-40V136a40 40 0 0140-40h152c22.09 0 48 17.91 48 40v40M368 336l80-80-80-80M176 256h256' stroke-linecap='round' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
              role="img"
              slot="icon-only"
            />
          </ion-button>
        </ion-buttons>
      </ion-toolbar>
    </ion-header>
    <ion-content
      class="main-content md"
      style="--offset-top: 0px; --offset-bottom: 0px;"
    >
      <ion-header
        class="md header-md header-collapse-condense"
        collapse="condense"
        role="banner"
      >
        <!---->
        <ion-toolbar
          class="toolbar-title-large md in-toolbar"
        >
          <ion-title
            class="md title-large"
            size="large"
          >
            About Tea Taster
          </ion-title>
        </ion-toolbar>
      </ion-header>
      <ion-list
        class="md list-md"
      >
        <ion-item
          class="item md item-fill-none in-list ion-focusable item-label"
        >
          <ion-label
            class="sc-ion-label-md-h sc-ion-label-md-s md"
          >
            Name
          </ion-label>
          <ion-note
            class="md"
            slot="end"
          >
            tea-taster
          </ion-note>
        </ion-item>
        <ion-item
          class="item md item-fill-none in-list ion-focusable item-label"
        >
          <ion-label
            class="sc-ion-label-md-h sc-ion-label-md-s md"
          >
            Description
          </ion-label>
          <ion-note
            class="md"
            slot="end"
          >
            Tea Tasting Notes
          </ion-note>
        </ion-item>
        <ion-item
          class="item md item-fill-none in-list ion-focusable item-label"
        >
          <ion-label
            class="sc-ion-label-md-h sc-ion-label-md-s md"
          >
            Version
          </ion-label>
          <ion-note
            class="md"
            slot="end"
          >
            2.0.0
          </ion-note>
        </ion-item>
        <ion-item
          class="item md item-fill-none in-list ion-focusable item-label"
        >
          <ion-label
            class="sc-ion-label-md-h sc-ion-label-md-s md"
          >
            Author
          </ion-label>
          <ion-note
            class="md"
            slot="end"
          >
            Ionic
          </ion-note>
        </ion-item>
      </ion-list>
    </ion-content>
  </div>
</DocumentFragment>
`;
